momepy/preprocessing.py

```diff
@@ -740,3 +745,368 @@ def _get_extrapolated_line(coords, tolerance, point=False):
+     to and from the new endpoints.
+
+    :param geometry: shapely.LineString, current edge
+    :param new_origin: shapely.Point or None, new origin point or None if origin remains the same
+    :param new_destination: shapely.Point or None, new destination point or None if destination remains the same
+
+    return list of edges in format (origin_id, destination_id, edge_data)
+    """
+    # Determine what endpoints were made into clusters:
+    edges_gdf["origin_cluster"] = edges_gdf[edge_from_col].apply(
+        lambda u: nodes_dict[u] if u in nodes_dict else -1
+    )
+    edges_gdf["destination_cluster"] = edges_gdf[edge_to_col].apply(
+        lambda v: nodes_dict[v] if v in nodes_dict else -1
+    )
+
+    # Determine what edges need to be simplified (either between diff. clusters or self-loops in a cluster):
+    edges_gdf.query(
+        "origin_cluster != destination_cluster or (('"
+        + edge_to_col
+        + "' == '"
+        + edge_from_col
+        + "') and origin_cluster >= 0)",
+        inplace=True,
+    )
+
+    # Determine the new point geometries (when exists):
+    cluster_df["geometry"] = gpd.points_from_xy(
+        cluster_df.x, cluster_df.y, crs=graph.graph["crs"]
+    )
+    edges_gdf["new_origin_pt"] = edges_gdf.origin_cluster.map(cluster_df.geometry, None)
+    edges_gdf["new_destination_pt"] = edges_gdf.destination_cluster.map(
+        cluster_df.geometry, None
+    )
+
+    # Determine the new geometry according to the simplification method:
+    if method == "linear":
+        edges_gdf["new_geometry"] = edges_gdf.apply(
+            lambda edge: _linear_simplification(
+                edge.geometry, edge.new_origin_pt, edge.new_destination_pt
+            ),
+            axis=1,
+        )
+    elif method == "crow_flies":
+        edges_gdf["new_geometry"] = edges_gdf.apply(
+            lambda edge: _crowflies_simplification(
+                edge.geometry, edge.new_origin_pt, edge.new_destination_pt
+            ),
+            axis=1,
+        )
+    elif method == "spider":
+        edges_gdf["new_geometry"] = edges_gdf.apply(
+            lambda edge: _spider_simplification(
+                edge.geometry,
+                edge.new_origin_pt,
+                edge.new_destination_pt,
+                1.5 * tolerance,
+            ),
+            axis=1,
+        )
+
+    # Rename and update the columns:
+    edges_gdf.reset_index(inplace=True)
+    edges_gdf.rename(
+        {
+            edge_from_col: "original_from",
+            edge_to_col: "original_to",
+            "origin_cluster": edge_from_col,
+            "destination_cluster": edge_to_col,
+            "geometry": "original_geometry",
+            "new_geometry": "geometry",
+        },
+        axis=1,
+        inplace=True,
+    )
+    edges_gdf.drop(columns=["new_origin_pt", "new_destination_pt"], inplace=True)
+    edges_gdf["length"] = edges_gdf["geometry"].apply(lambda x: x.length)
```

Comment:
```suggestion
    edges_gdf["length"] = edges_gdf.length
```
This will use vectorised operation instead of a loop.

Reply:
Btw, why do you need this anyway?

Reply:
Updates length parameter of rebuilt edges to match their new geometries.